src/group_theory/finiteness.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import data.set.finite
+import group_theory.submonoid.operations
+
+/-!
+# Finitely generated monoid.
+
+We define finitely generated monoids.
+
+## Main definition
+
+* `monoid.fg`, `add_monoid.fg` : A monoid is finitely generated.
+
+-/
+
+section monoid
+
+variables {M N : Type*} [monoid M] [add_monoid N]
+
+/-- A submonoid of `M` is finitely generated if it is the closure of a finite subset of `M`. -/
+@[to_additive]
+def submonoid.fg (P : submonoid M) : Prop := ∃ S : finset M, submonoid.closure ↑S = P
+
+/-- An additive submonoid of `N` is finitely generated if it is the closure of a finite subset of
+`M`. -/
+add_decl_doc add_submonoid.fg
+
+variables (M N)
+
+/-- A monoid is finitely generated if it is finitely generated as a submonoid of itself. -/
+class monoid.fg : Prop := (out : (⊤ : submonoid M).fg)
+
+/-- An additive monoid is finitely generated if it is finitely generated as an additive submonoid of
+itself. -/
+class add_monoid.fg : Prop := (out : (⊤ : add_submonoid N).fg)
```

Comment:
```suggestion

class add_monoid.fg : Prop := (out : (⊤ : add_submonoid N).fg)

attribute [to_additive] monoid.fg
```

I think this file would read more cleanly though if you put all the `submonoid.fg` stuff at the top and all the uninteresting `monoid.fg` glue at the bottom.

Reply:
Just checking you saw the second half of my comment - I think it would be easier to follow if the lemmas didn't alternate between `submonoid` and `monoid`.

Reply:
Ah, nevermind, I see you did.